learn/how-tos/user-driven-dynamic-roles.md

```diff
@@ -0,0 +1,103 @@
+---
+title: "User-Driven Dynamic Roles"
+id: ""
+---
+---
+
+Role-based Access Control (RBAC) is essential to restrict end-users from accessing specific sources and actions, including Pages, Widgets, and Services. It is easy to manage application roles for small and single-page applications. However, complex applications can have over a hundred permissions, which is not quite simple. 
+
+## Key RBAC Challenges
+
+1. Generally, the roles are managed and assigned during the application design time on WaveMaker.
+2. To add a new role to the application, the code needs to be ch  anged and re-deploy.
+3. For instance, when working with large Enterprise applications, app administrators typically have to manage multiple app users and roles.
+
+### How this document helps 
+
+1. Handling complex application permissions by defining application roles dynamically without many code changes.
+2. Helps developers design and manage the roles and allow role-based actions dynamically with a Proof of Concept (PoC).
+
+## Implementation
+
+To enable this approach in a WaveMaker application, add the following tables to the database.
+
+### Add Tables to the DB Schema 
+
+ 1. Permissions - To hold the WaveMaker defined permissions
+ 2. Roles - To define the custom roles
+ 3. Rolemapping - Association between permissions and roles
+ 4. UserRoleMapping - Association between users and roles
+
+[![pp_run](/learn/assets/dynamic-roles-1.png)](/learn/assets/dynamic-roles-1.png)
+
+### Define WaveMaker Security
+
+1. Go to Security → App Roles. Add the Permissions
+
+[![pp_run](/learn/assets/dynamic-roles-security-1.png)](/learn/assets/dynamic-roles-security-1.png) 
+
+2. Go to Security → Security Providers → Role Mapping section. Make sure to add the query according to the requirement. 
+
+[![pp_run](/learn/assets/dynamic-roles-security-2.png)](/learn/assets/dynamic-roles-security-2.png) 
+ 
+3. Assign permissions to Pages and Services. 
+
+
+## Use Case
+
+### Indentifying Problem
+
+User roles on WaveMaker are tightly coupled with the Studio. It is not possible to add or remove permissions during runtime.
+
+### Proposed Solution
+
+Consider app roles on WaveMaker security as User Permissions (P1, P2) have an additional layer of roles (Role-A, Role-B). 
```

Comment:
Not clear. Identify the feature of WM you are using to enable dynamic roles? Are you talking about setting the role on loggedinuser instance?